### ISSUE TITLE
Newly attached disk is wrongly detached by already ongoing `detachDisk` operation in GCP

### ISSUE THREAD

**Issue by ialidzhikov** [kubernetes-sigs/gcp-compute-persistent-disk-csi-driver#965]
/sig storage
/kind bug

**What happened**:
A `ControllerUnpublishVolume` call that for example fails with `failed to poll the operation` or `context deadline excceeded` actually succeeds in the background after 1-2 minutes and wrongly detaches already newly attached disk.

**What you expected to happen**:
`ControllerUnpublishVolume` to be idempotent and to do not have any operations in GCP that detach the disk when it is is newly attached.

**How to reproduce it (as minimally and precisely as possible)**:

1. Create a single Node GCP cluster with K8s 1.20.13.

2. Create a StorageClass and 40 StatefulSets.

   <details>
   
   ```yaml
   allowVolumeExpansion: true
   apiVersion: storage.k8s.io/v1
   kind: StorageClass
   metadata:
     name: pd-ssd
   parameters:
     type: pd-ssd
   provisioner: kubernetes.io/gce-pd
   reclaimPolicy: Delete
   volumeBindingMode: Immediate
   ```

   ```bash
   #!/bin/bash
   
   set -o errexit
   set -o nounset
   set -o pipefail
   
   for i in {1..40}
   do
       cat <<EOF | kubectl create -f -
   apiVersion: v1
   kind: Service
   metadata:
     name: app$i
     labels:
       app: app$i
   spec:
     ports:
     - port: 80
       name: web
     clusterIP: None
     selector:
       app: app$i
   ---
   apiVersion: apps/v1
   kind: StatefulSet
   metadata:
     name: app$i
   spec:
     serviceName: app$i
     replicas: 1
     selector:
       matchLabels:
         app: app$i
     template:
       metadata:
         labels:
           app: app$i
       spec:
         containers:
           - name: app$i
             image: centos
             command: ["/bin/sh"]
             args: ["-c", "while true; do echo app$i $(date -u) >> /data/out.txt; sleep 5; done"]
             volumeMounts:
             - name: persistent-storage-app$i
               mountPath: /data
     volumeClaimTemplates:
     - metadata:
         name: persistent-storage-app$i
       spec:
         storageClassName: pd-ssd
         accessModes: [ "ReadWriteOnce" ]
         resources:
           requests:
             storage: 1Gi
   EOF
   
       sleep 10s
   done   
   ```

   </details>

4. Rollout all StatefulSets at once.

   ```
   $ k rollout restart sts
   ```

5. Make sure that after the rollout on the Node some of the attached devices are not mounted (despite that all Pods are reported to be running)

   <details>

   ```
   # lsblk
   NAME MAJ:MIN RM  SIZE RO TYPE MOUNTPOINTS
   sda    8:0    0   30G  0 disk
   |-sda1
   |      8:1    0  128M  0 part /boot/efi
   |-sda2
   |      8:2    0    1G  0 part /usr
   `-sda3
          8:3    0 28.9G  0 part /
   sdb    8:16   0    1G  0 disk /var/lib/kubelet/pods/67757f27-326b-46a9-950a-d62a3affa3f8/volumes/kubernetes.io~csi/pv--8b056cb3-c991-4308-959b-682a0253488d/mount
                                 /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv--8b056cb3-c991-4308-959b-682a0253488d/globalmount
   sdc    8:32   0    1G  0 disk /var/lib/kubelet/pods/9562af2b-2106-48de-b4c6-d09447a8e8bc/volumes/kubernetes.io~csi/pv--51817eac-7469-49c0-8611-bf743666917b/mount
                                 /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv--51817eac-7469-49c0-8611-bf743666917b/globalmount
   sdd    8:48   0    1G  0 disk /var/lib/kubelet/pods/ccdc07a1-f039-455a-8329-c809a5597fd4/volumes/kubernetes.io~csi/pv--7312b907-c3f3-40cd-96a4-17b7c93434a7/mount
                                 /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv--7312b907-c3f3-40cd-96a4-17b7c93434a7/globalmount
   sde    8:64   0    1G  0 disk /var/lib/kubelet/pods/95cab30c-aa16-4793-b78d-33a1bca1c8a0/volumes/kubernetes.io~csi/pv--b5cb0051-664c-41f7-af9f-a143a117c3d9/mount
                                 /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv--b5cb0051-664c-41f7-af9f-a143a117c3d9/globalmount
   sdf    8:80   0    1G  0 disk /var/lib/kubelet/pods/105b5b6a-9cde-4208-ae38-18fdad4ec00b/volumes/kubernetes.io~csi/pv--dc7e1a73-df94-4618-bca9-872f04022b47/mount
                                 /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv--dc7e1a73-df94-4618-bca9-872f04022b47/globalmount
   sdg    8:96   0    1G  0 disk /var/lib/kubelet/pods/7a74d15c-965c-4abd-8d69-7d02073872a4/volumes/kubernetes.io~csi/pv--06aa6718-c297-4501-898e-ba75f3abb072/mount
                                 /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv--06aa6718-c297-4501-898e-ba75f3abb072/globalmount
   sdh    8:112  0    1G  0 disk /var/lib/kubelet/pods/008cd489-ac3c-41e9-97f3-748af62c2952/volumes/kubernetes.io~csi/pv--0a567d3b-c20f-4f99-9f4e-83441e4dda90/mount
                                 /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv--0a567d3b-c20f-4f99-9f4e-83441e4dda90/globalmount
   sdj    8:144  0    1G  0 disk /var/lib/kubelet/pods/d4256a6a-3786-4c1e-ae99-d900ca029261/volumes/kubernetes.io~csi/pv--a856c27d-3003-494e-81c9-ec7421c92af1/mount
                                 /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv--a856c27d-3003-494e-81c9-ec7421c92af1/globalmount
   sdk    8:160  0    1G  0 disk /var/lib/kubelet/pods/0ccb8da5-d885-4518-b6b6-581575353354/volumes/kubernetes.io~csi/pv--834611cc-b6b9-4bee-8949-2f58095b5fd0/mount
                                 /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv--834611cc-b6b9-4bee-8949-2f58095b5fd0/globalmount
   sdl    8:176  0    1G  0 disk /var/lib/kubelet/pods/8d59ced6-e6f1-4427-94f3-adb4c5805149/volumes/kubernetes.io~csi/pv--d4cd30c2-631f-4dcd-a9ee-13ec9da9b911/mount
                                 /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv--d4cd30c2-631f-4dcd-a9ee-13ec9da9b911/globalmount
   sdm    8:192  0    1G  0 disk /var/lib/kubelet/pods/33e786d7-248d-496b-8c41-588630c4c952/volumes/kubernetes.io~csi/pv--1ebf01e1-7b09-4652-948c-6596fb3d3404/mount
                                 /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv--1ebf01e1-7b09-4652-948c-6596fb3d3404/globalmount
   sdn    8:208  0    1G  0 disk /var/lib/kubelet/pods/66c5d8db-fd17-420d-8f12-e62faf27cd9a/volumes/kubernetes.io~csi/pv--4c6fc335-e10b-4fa1-9a75-a63804e3748a/mount
                                 /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv--4c6fc335-e10b-4fa1-9a75-a63804e3748a/globalmount
   sdo    8:224  0    1G  0 disk /var/lib/kubelet/pods/b43e379f-da8b-41c7-a3c7-e0d2aef17f74/volumes/kubernetes.io~csi/pv--7979a791-9755-4e35-8e96-929b4905f263/mount
                                 /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv--7979a791-9755-4e35-8e96-929b4905f263/globalmount
   sdt   65:48   0    1G  0 disk /var/lib/kubelet/pods/554a4187-0986-457f-a37a-b07e84ca4d66/volumes/kubernetes.io~csi/pv--8585ce75-6b39-4f85-ae74-b9eed193e955/mount
                                 /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv--8585ce75-6b39-4f85-ae74-b9eed193e955/globalmount
   sdw   65:96   0    1G  0 disk /var/lib/kubelet/pods/ff6ce1a7-8a13-43ab-a5f9-d5fdfe204203/volumes/kubernetes.io~csi/pv--eaa26743-23b6-4f22-9949-bfcf2e12cf1d/mount
                                 /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv--eaa26743-23b6-4f22-9949-bfcf2e12cf1d/globalmount
   sdy   65:128  0    1G  0 disk
   sdz   65:144  0    1G  0 disk /var/lib/kubelet/pods/5481cd0e-5aa2-493f-a906-daabcccfef1e/volumes/kubernetes.io~csi/pv--105d7e8b-0b76-48b0-a0a6-9b58f18c8dd8/mount
                                 /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv--105d7e8b-0b76-48b0-a0a6-9b58f18c8dd8/globalmount
   sdaa  65:160  0    1G  0 disk /var/lib/kubelet/pods/8b2ac03f-6a4b-4693-87f5-e2018c9e4382/volumes/kubernetes.io~csi/pv--28d31cb5-d68a-4c5c-8e30-c09df4722b25/mount
                                 /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv--28d31cb5-d68a-4c5c-8e30-c09df4722b25/globalmount
   sdac  65:192  0    1G  0 disk /var/lib/kubelet/pods/4f4c2cfe-c3a1-4117-bc4f-2619d4ce8de7/volumes/kubernetes.io~csi/pv--63591fe1-e7d1-4781-a0fc-bd53a3f8095d/mount
                                 /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv--63591fe1-e7d1-4781-a0fc-bd53a3f8095d/globalmount
   sdad  65:208  0    1G  0 disk /var/lib/kubelet/pods/f2a46880-270a-44ad-b258-6867e2678e12/volumes/kubernetes.io~csi/pv--153a4195-2b29-4b59-a21e-f006bdaf5841/mount
                                 /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv--153a4195-2b29-4b59-a21e-f006bdaf5841/globalmount
   sdae  65:224  0    1G  0 disk /var/lib/kubelet/pods/d5f28c30-39c6-4bfd-9079-e4beec1e5a43/volumes/kubernetes.io~csi/pv--2f29fa37-51b0-4cda-bfce-8d4f0d4e7bcc/mount
                                 /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv--2f29fa37-51b0-4cda-bfce-8d4f0d4e7bcc/globalmount
   sdaf  65:240  0    1G  0 disk
   sdag  66:0    0    1G  0 disk /var/lib/kubelet/pods/34e93c32-e53e-4a35-b149-880bb7b70237/volumes/kubernetes.io~csi/pv--1a1d88c4-68fc-44ea-8c06-9c9ef95d985e/mount
                                 /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv--1a1d88c4-68fc-44ea-8c06-9c9ef95d985e/globalmount
   sdak  66:64   0    1G  0 disk
   sdal  66:80   0    1G  0 disk /var/lib/kubelet/pods/3f245461-a49e-411d-982b-827199e96979/volumes/kubernetes.io~csi/pv--a360d97a-c2d6-40bd-87ea-85f8049e2ab2/mount
                                 /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv--a360d97a-c2d6-40bd-87ea-85f8049e2ab2/globalmount
   sdam  66:96   0    1G  0 disk /var/lib/kubelet/pods/1d635bb0-4efa-4091-9a2e-97509b63fcc8/volumes/kubernetes.io~csi/pv--b8af17c7-1f1c-416e-b3d8-be8e64f5fa79/mount
                                 /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv--b8af17c7-1f1c-416e-b3d8-be8e64f5fa79/globalmount
   sdan  66:112  0    1G  0 disk
   sdao  66:128  0    1G  0 disk
   sdap  66:144  0    1G  0 disk /var/lib/kubelet/pods/24eb8e89-905e-4763-bac2-89542b4fc5ee/volumes/kubernetes.io~csi/pv--13c329ab-29cb-4c98-bf16-1d9434636688/mount
                                 /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv--13c329ab-29cb-4c98-bf16-1d9434636688/globalmount
   sdaq  66:160  0    1G  0 disk /var/lib/kubelet/pods/ca37a2f4-0520-47c4-9fce-0081fa0446fe/volumes/kubernetes.io~csi/pv--21f1b6d7-e61f-44b3-a287-33e2fb73031c/mount
                                 /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv--21f1b6d7-e61f-44b3-a287-33e2fb73031c/globalmount
   sdar  66:176  0    1G  0 disk
   sdas  66:192  0    1G  0 disk
   sdat  66:208  0    1G  0 disk
   sdau  66:224  0    1G  0 disk
   sdav  66:240  0    1G  0 disk /var/lib/kubelet/pods/fc366e7b-1738-4a72-9ca1-70daf2d9e757/volumes/kubernetes.io~csi/pv--0b1e4ea6-fef7-4ce5-a02b-604462b7f225/mount
                                 /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv--0b1e4ea6-fef7-4ce5-a02b-604462b7f225/globalmount
   sdaw  67:0    0    1G  0 disk
   sdax  67:16   0    1G  0 disk /var/lib/kubelet/pods/6cda5288-33dd-47e6-a62e-e11d81490143/volumes/kubernetes.io~csi/pv--b4717f5a-a78d-4a8e-a1db-bd4d9db7d7e7/mount
                                 /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv--b4717f5a-a78d-4a8e-a1db-bd4d9db7d7e7/globalmount
   sday  67:32   0    1G  0 disk
   sdaz  67:48   0    1G  0 disk /var/lib/kubelet/pods/eb4450db-9c09-490b-91d3-487076fcc26c/volumes/kubernetes.io~csi/pv--9528b812-b4cf-41cc-9c3b-3d91a999b861/mount
                                 /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv--9528b812-b4cf-41cc-9c3b-3d91a999b861/globalmount
   sdba  67:64   0    1G  0 disk
   ```

   </details>
   
   In the output above these are `sdy`, `sdaf`, `sdak`, `sdan`, `sdao`, `sdar`, `sdas`, `sdat`, `sdau`, `sdaw`, `sday` and `sdba`.
   `for i in {1..40}; do k exec app$i-0 -- tail -n 1 /data/out.txt; done` reports that 12 Pods have IO errors (12 devices that are not mounted and 12 Pods that have IO errors):

   <details>

   ```
   app1 Tue Apr 26 07:00:19 UTC 2022
   app2 Tue Apr 26 07:00:26 UTC 2022
   tail: cannot open '/data/out.txt' for reading: Input/output error
   command terminated with exit code 1
   app4 Tue Apr 26 07:00:37 UTC 2022
   app5 Tue Apr 26 07:00:42 UTC 2022
   app6 Tue Apr 26 07:00:48 UTC 2022
   app7 Tue Apr 26 07:00:53 UTC 2022
   tail: cannot open '/data/out.txt' for reading: Input/output error
   command terminated with exit code 1
   app9 Tue Apr 26 07:01:04 UTC 2022
   app10 Tue Apr 26 07:01:10 UTC 2022
   app11 Tue Apr 26 07:01:15 UTC 2022
   app12 Tue Apr 26 07:01:21 UTC 2022
   tail: cannot open '/data/out.txt' for reading: Input/output error
   command terminated with exit code 1
   app14 Tue Apr 26 07:01:32 UTC 2022
   tail: cannot open '/data/out.txt' for reading: Input/output error
   command terminated with exit code 1
   tail: cannot open '/data/out.txt' for reading: Input/output error
   command terminated with exit code 1
   app17 Tue Apr 26 07:01:48 UTC 2022
   app18 Tue Apr 26 07:01:53 UTC 2022
   app19 Tue Apr 26 07:01:59 UTC 2022
   tail: cannot open '/data/out.txt' for reading: Input/output error
   command terminated with exit code 1
   app21 Tue Apr 26 07:02:10 UTC 2022
   app22 Tue Apr 26 07:02:15 UTC 2022
   app23 Tue Apr 26 07:02:21 UTC 2022
   app24 Tue Apr 26 07:02:26 UTC 2022
   tail: cannot open '/data/out.txt' for reading: Input/output error
   command terminated with exit code 1
   app26 Tue Apr 26 07:02:37 UTC 2022
   app27 Tue Apr 26 07:02:43 UTC 2022
   app28 Tue Apr 26 07:02:48 UTC 2022
   app29 Tue Apr 26 07:02:54 UTC 2022
   app30 Tue Apr 26 07:02:59 UTC 2022
   tail: cannot open '/data/out.txt' for reading: Input/output error
   command terminated with exit code 1
   app32 Tue Apr 26 07:03:10 UTC 2022
   tail: cannot open '/data/out.txt' for reading: Input/output error
   command terminated with exit code 1
   app34 Tue Apr 26 07:03:21 UTC 2022
   tail: cannot open '/data/out.txt' for reading: Input/output error
   command terminated with exit code 1
   app36 Tue Apr 26 07:03:32 UTC 2022
   app37 Tue Apr 26 07:03:37 UTC 2022
   tail: cannot open '/data/out.txt' for reading: Input/output error
   command terminated with exit code 1
   tail: cannot open '/data/out.txt' for reading: Input/output error
   command terminated with exit code 1
   app40 Tue Apr 26 07:03:54 UTC 2022
   ```

   </details>

   The Node events:
   ```
   Events:
   Type     Reason                   Age                    From            Message
   ----     ------                   ----                   ----            -------
   Normal   NodeHasNoDiskPressure    21m (x8 over 21m)      kubelet         Node worker-1-z1-67fbc-4jn6k status is now: NodeHasNoDiskPressure
   Normal   NodeHasSufficientMemory  21m (x8 over 21m)      kubelet         Node worker-1-z1-67fbc-4jn6k status is now: NodeHasSufficientMemory
   Normal   Starting                 21m                    kube-proxy      Starting kube-proxy.
   Normal   Starting                 15m                    kube-proxy      Starting kube-proxy.
   Warning  IOError                  10m                    kernel-monitor  Buffer I/O error on dev sdp, logical block 145, lost async page write
   Warning  Ext4Warning              10m                    kernel-monitor  EXT4-fs warning (device sdp): ext4_end_bio:345: I/O error 10 writing to inode 12 starting block 32898)
   Warning  IOError                  10m                    kernel-monitor  Buffer I/O error on device sdp, logical block 32897
   Warning  IOError                  10m                    kernel-monitor  Buffer I/O error on dev sdp, logical block 131072, lost sync page write
   Warning  IOError                  9m58s                  kernel-monitor  Buffer I/O error on device sdu, logical block 32897
   Warning  IOError                  9m58s                  kernel-monitor  Buffer I/O error on dev sdu, logical block 131072, lost sync page write
   Warning  Ext4Warning              9m58s                  kernel-monitor  EXT4-fs warning (device sdu): ext4_end_bio:345: I/O error 10 writing to inode 12 starting block 32898)
   Warning  Ext4Warning              9m50s                  kernel-monitor  EXT4-fs warning (device sdah): ext4_end_bio:345: I/O error 10 writing to inode 12 starting block 32898)
   Warning  IOError                  9m50s                  kernel-monitor  Buffer I/O error on device sdah, logical block 32897
   Warning  IOError                  9m50s                  kernel-monitor  Buffer I/O error on dev sdah, logical block 0, lost async page write
   Warning  IOError                  9m50s                  kernel-monitor  Buffer I/O error on dev sdah, logical block 131072, lost sync page write
   Warning  Ext4Error                9m37s (x3 over 9m47s)  kernel-monitor  EXT4-fs error (device sdah): __ext4_find_entry:1534: inode #2: comm sh: reading directory lblock 0
   Warning  Ext4Error                9m34s (x7 over 10m)    kernel-monitor  EXT4-fs error (device sdp): __ext4_find_entry:1534: inode #2: comm sh: reading directory lblock 0
   Warning  Ext4Error                5m7s (x46 over 9m53s)  kernel-monitor  EXT4-fs error (device sdu): __ext4_find_entry:1534: inode #2: comm sh: reading directory lblock 0
   Warning  Ext4Error                7s (x42 over 9m18s)    kernel-monitor  EXT4-fs error (device sdai): __ext4_find_entry:1534: inode #2: comm sh: reading directory lblock 0
   ```

**Anything else we need to know?**:
We see that there are discrepancies between the attached devices and the mount table:
```
$ mount | grep pv--661ff519-cc27-483a-908c-ba07293b3f9b
/dev/sdm on /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv--661ff519-cc27-483a-908c-ba07293b3f9b/globalmount type ext4 (rw,relatime)
/dev/sdm on /var/lib/kubelet/pods/e878681e-9572-4730-bae5-4290c6fe31c1/volumes/kubernetes.io~csi/pv--661ff519-cc27-483a-908c-ba07293b3f9b/mount type ext4 (rw,relatime)

# ls -la /dev/disk/by-id/ | grep pv--661ff519-cc27-483a-908c-ba07293b3f9b
lrwxrwxrwx 1 root root   10 Apr 27 08:47 google-pv--661ff519-cc27-483a-908c-ba07293b3f9b -> ../../sdas
lrwxrwxrwx 1 root root   10 Apr 27 08:47 scsi-0Google_PersistentDisk_pv--661ff519-cc27-483a-908c-ba07293b3f9b -> ../../sdas
```

For the PV the mount table reports that it is using device `/dev/sdm` but `/dev/disk/by-id/` reports that the PV is attached to `/dev/sdas`.

After long investigation we see that the problem is caused by a detachDisk request that completes after a long period of time and wrongly detaches the PV that is attached to the Node.

The relevant csi-driver controller logs are:
```
I0427 08:44:32.006569       1 utils.go:67] /csi.v1.Controller/ControllerUnpublishVolume called with request: volume_id:"projects/UNSPECIFIED/zones/europe-west2-b/disks/pv--661ff519-cc27-483a-908c-ba07293b3f9b" node_id:"projects/foo/zones/europe-west2-b/instances/worker-1-z1-5c446-85trq"
E0427 08:44:47.748537       1 gce-compute.go:698] WaitForOp(op: operation-1651049072368-5dd9ed1771d5f-84095ef1-8eb45af4, zone: "europe-west2-b") failed to poll the operation
I0427 08:44:47.862202       1 utils.go:67] /csi.v1.Controller/ControllerUnpublishVolume called with request: volume_id:"projects/UNSPECIFIED/zones/europe-west2-b/disks/pv--661ff519-cc27-483a-908c-ba07293b3f9b" node_id:"projects/foo/zones/europe-west2-b/instances/worker-1-z1-5c446-85trq" 
I0427 08:44:55.061813       1 controller.go:449] ControllerUnpublishVolume succeeded for disk Key{"pv--661ff519-cc27-483a-908c-ba07293b3f9b", zone: "europe-west2-b"} from node projects/foo/zones/europe-west2-b/instances/worker-1-z1-5c446-85trq
I0427 08:45:10.462657       1 utils.go:67] /csi.v1.Controller/ControllerPublishVolume called with request: volume_id:"projects/UNSPECIFIED/zones/europe-west2-b/disks/pv--661ff519-cc27-483a-908c-ba07293b3f9b" node_id:"projects/foo/zones/europe-west2-b/instances/worker-1-z1-5c446-85trq" volume_capability:<mount:<fs_type:"ext4" > access_mode:<mode:SINGLE_NODE_WRITER > > volume_context:<key:"partition" value:"" > 
I0427 08:45:25.662586       1 utils.go:67] /csi.v1.Controller/ControllerPublishVolume called with request: volume_id:"projects/UNSPECIFIED/zones/europe-west2-b/disks/pv--661ff519-cc27-483a-908c-ba07293b3f9b" node_id:"projects/foo/zones/europe-west2-b/instances/worker-1-z1-5c446-85trq" volume_capability:<mount:<fs_type:"ext4" > access_mode:<mode:SINGLE_NODE_WRITER > > volume_context:<key:"partition" value:"" > 
I0427 08:45:26.071681       1 controller.go:364] ControllerPublishVolume succeeded for disk Key{"pv--661ff519-cc27-483a-908c-ba07293b3f9b", zone: "europe-west2-b"} to instance projects/foo/zones/europe-west2-b/instances/worker-1-z1-5c446-85trq, already attached.
I0427 08:46:55.161344       1 utils.go:67] /csi.v1.Controller/ControllerPublishVolume called with request: volume_id:"projects/UNSPECIFIED/zones/europe-west2-b/disks/pv--661ff519-cc27-483a-908c-ba07293b3f9b" node_id:"projects/foo/zones/europe-west2-b/instances/worker-1-z1-5c446-85trq" volume_capability:<mount:<fs_type:"ext4" > access_mode:<mode:SINGLE_NODE_WRITER > > volume_context:<key:"partition" value:"" > 
```

From the audit logs for this PV we see the following sequence of events:
1. (`11:44:32.453`) Operation 1 (`operation-1651049072368-5dd9ed1771d5f-84095ef1-8eb45af4`) `detachDisk` starts. As you can see above this operation fails on csi-driver side with `failed to poll the operation`.
2. (`11:44:48.174`) Operation 2 (`operation-1651049088072-5dd9ed266bc15-3aeb7659-089d253f`) `detachDisk` starts.
3. (`11:44:53.804`) Operation 2 (`operation-1651049088072-5dd9ed266bc15-3aeb7659-089d253f`) `detachDisk` succeeds.
4. (`11:45:11.153`) Operation 3 (`operation-1651049111074-5dd9ed3c5b6b5-28b448f4-7ac81e74`) `attachDisk` starts
5. (`11:45:23.071`) Operation 3 (`operation-1651049111074-5dd9ed3c5b6b5-28b448f4-7ac81e74`) `attachDisk` succeeds
6. (`11:46:07.751`) Operation 1 (`operation-1651049072368-5dd9ed1771d5f-84095ef1-8eb45af4`) succeeds and wrongly detaches the already attached disk.

Obviously step 6. is quite wrong and a detach request from 1m30s ago completes now and wrongly detaches already the newly attached disk.

**Environment**:

- Kubernetes version (use `kubectl version`): 1.20.13
- Cloud provider or hardware configuration:
- Others:
   `gcp-csi-driver` - `v1.3.4-gke.0`

**Comment by mattcary** [comment]
Yes, it's exactly that GCP queuing behavior that this is fixing. It turns out on GCP, there is a per {disk,instance} FILO queue for operations. So if operations are queued up, attach and detach operations will get issued out of order, causing the behavior you see.

**Comment by dguendisch** [comment]
So #923 in its current form should already be enough for non multi-master setups to solve the issue described here, correct?

**Comment by mattcary** [comment]
> Btw, is this issue affecting all existing csi-driver versions or is it specific to a certain version range? 

This affects all versions (as you pointed out, the problem is really with the unexpected behavior of GCP queuing that happens when lots of attach/detach operations are going on). It's possible however that recent changes (either to our driver, like the backoff for failed operations, or to k8s, like the new uncertain volume status) have made it more likely to hit this GCP queuing case. I haven't thought about that in detail TBH, we're just focusing on fixing the issue :-)

> Do you also recommend update to the latest v1.5.1? Is there any know bugfix that would save us headaches?

We just updated GKE to v1.5.1 if that answers your question :-). It has several bugfixes in it.

**Comment by mattcary** [comment]
This should be fixed with the 1.7 release.

/close
